text/0000-execution-environments.md

```diff
@@ -0,0 +1,225 @@
+- What is the test result format like [TAP](https://en.wikipedia.org/wiki/Test_Anything_Protocol)?
+- How to process the test results
+- What to do with the results
+
+This narrows the scope of the Buildpack responsibilities to what it's already good at:
+
+- Produce Test Environment
+- How to launch the process
+  - launch.toml with test process marked as default (recommendation)
+
+## Setting the Execution Enviroment for Build + Buildpack
+A platform will set the `CNB_EXEC_ENV` env var to the execution environment desired. Buildpacks can than read this env var to branch or switch on logic needed based on the execution environment.
+
+In addition, Builder Authors, Buildpack Authors, and App Developers will be able to specify various options to a specific execution enviroment using the `exec-env` key.
+
+## Development Environments
+The specifics of creating development enviroments are out of scope of this RFC, but it's not hard to extrapolate how these kind of changes can assist in creating Buildpacks for development environments.
+
+# How it Works
+[how-it-works]: #how-it-works
+
+## `exec-env` key in TOML
+
+In order to support additional execution environments an `exec-env` key will be added to various TOML tables in the project. The value can be any string with `all` having special meaning. `all` will apply to all execution environments and will be the default if not specified. This should make it backwards compatible and optional. When `exec-env` is not set to `all`, the table settings will only be applied to that execution environment.
+
+### Project Descriptor - `project.toml` (App Developers)
+
+This file be extended by adding `exec-env` to the following tables:
+
+`[[io.buildpacks.group]]`
+`[[io.buildpacks.pre.group]]`
+`[[io.buildpacks.post.group.env]]`
+`[[io.buildpacks.build.env]]`
+
+An example would look like this:
+
+```TOML
+[_]
+schema-version = "0.3"
+
+[[io.buildpacks.group]]
+id = "buildpacks/ruby"
+version = "latest"
+
+[[io.buildpacks.group]]
+id = "buildpacks/metrics-agent"
+version = "latest"
+exec-env = "production"
+
+[[io.buildpacks.group]]
+id = "buildpacks/headless-chrome"
+version = "latest"
+exec-env = "test"
+
+[[io.buildpacks.post.group]]
+id = "buildpacks/procfile"
+version = "latest"
+
+[[io.buildpacks.buildenv]]
+name = "RAILS_ENV"
+value = "production"
+exec-env = "production"
+
+[[io.buildpacks.build.env]]
+name = "RAILS_ENV"
+value = "test"
+exec-env = "test"
+
+[[io.buildpacks.build.env]]
+name = "PARALLEL_WORKERS"
+value = "4"
+exec-env = "test"
+```
+
+### `builder.toml` (Builder Authors)
+
+The only table that `exec-env` will be added to is `[[order.group.env]]`.
+
+### `buildpack.toml` (Buildpack Authors)
+
+The only table that `exec-env` will be added to is `[[buildpack.order.group]]`. This only is applicable for composite buildpacks.
+
+## `CNB_EXEC_ENV` Environment Variable
+
+This env var will reserve the following values:
+
+* production
+* test
+* development
+
+### Buildpack API
```

Comment:
Can we perhaps get a conceptual example of what a buildpack author might do with this new execution environment information?

I have in my head a language buildpack author may...
- install test group dependencies
- create `ruby-tests` process / `ruby-tests-verbose` process
- Skip cleaning up things they might clean up otherwise for production images
- Set env vars to test mode operation (RAILS_ENV or similar)